Graph colouring cluster ids

mundya opened the pull request:
Cluster IDs are used to ensure the multicast nets representing signals which originate from multiple chips can be differentiated if they take different routes at the same router. Previously each chip was assigned a unique identifier but this is unnecessary and wastes both routing table entries and keyspace.

This PR uses graph colouring to reduce the number of different cluster IDs applied, with the result the fewer bits in the key and fewer routing tables are used to simulate the same model.

mundya committed:
Improve storage of vertices in the netlist

Store a map from operator to vertices rather than just a list of
vertices. This makes things a little clearer in places and will make
later assignment of cluster IDs cleaner.

mundya committed:
Assign cluster IDs using graph colouring

Cluster IDs are used to ensure the multicast nets representing signals
which originate from multiple chips can be differentiated if they take
different routes at the same router. Previously each chip was assigned a
unique identifier but this is unnecessary and wastes both routing table
entries and keyspace. This commit uses the graph colouring to try to
reduce the number of different cluster IDs applied, with the result the
fewer bits in the key and fewer routing tables are used to simulate the
same model.

mundya commented on nengo_spinnaker/netlist/key_allocation.py:
```diff
@@ -106,26 +106,163 @@ def build_mn_net_graph(nets_routes, prior_constraints=None):
+            for vertex in vertices:
+                vertex.cluster = 0
+        else:
+            # Otherwise try to allocate a few cluster IDs as are required to
```
"Otherwise try to allocate _as_ few"

mundya commented on nengo_spinnaker/netlist/key_allocation.py:
```diff
@@ -106,26 +106,163 @@ def build_mn_net_graph(nets_routes, prior_constraints=None):
+                chips_signals_routes_clusters[chip][signal][rs].add(source)
+
+    # Build the graph representing the constraints on cluster identities by
+    # iterating over this new data.
```
"~~by iterating...~~"

neworderofjamie commented on nengo_spinnaker/builder/builder.py:
```diff
@@ -340,8 +339,10 @@ def make_netlist(self, *args, **kwargs):
 
-            operator_vertices[op] = vxs
-            vertices.append(vxs)
+            if isinstance(vxs, (list, tuple, set)):
```
This is unpleasant - could operators whose `make_vertices` method returns a single vertex do so wrapped in a tuple?

neworderofjamie reviewed (commented):
Looks good to me - nice re-use of colour_graph

mundya committed:
`make_vertices` returns iterable of vertices

Require that `make_vertices` always returns an iterable of vertices. As
suggested by @neworderofjamie.
